src/common/string-builder.ts

```diff
@@ -81,8 +82,39 @@ class StringBuilder {
         return this.printQueue.join(' ');
     }
 
-   combineURLs(baseURL: string, relativeURL?: string): string {
-        return relativeURL ? new URL(relativeURL, baseURL || undefined).toString() : baseURL;
+    combineURLs(baseURL: string, relativeURL?: string): string {
+        const createURL = (url: string) => {
+            try {
+                return new URL(url)
+            } catch (e) {
+                return null
+            }
+        }
+
+        let url: string = '';
+        const isFirstUrl = (baseURL !== '' && baseURL !== undefined);
+        const firstURL = isFirstUrl ? new URL(baseURL) : null;
+        const isRelativeUrl = relativeURL && relativeURL !== '';
+        const isRelativeUrlHaveHost = relativeURL && createURL(relativeURL) ? true : false;
+        const isBaseUrlHaveSubpath = firstURL && firstURL.pathname !== '' ? true : false;
+        
+        if (!isRelativeUrl) {
+            return baseURL;
+        }
+
+        if (isRelativeUrlHaveHost) {
+            return relativeURL as string;
+        }
+
+        if (!isRelativeUrlHaveHost) {
+            if (isBaseUrlHaveSubpath && firstURL) {
+                url = (new URL(path.join(firstURL.pathname, relativeURL as string), baseURL)).toString();
+            } else {
+                url = (new URL(relativeURL as string, baseURL)).toString();
+            }
+        }
+
+        return url;
```

Comment:
If you refer to the code below, I think it can be written more concisely.

```ts
function combineURLs(baseURL: string, relativeURL?: string): string {
    if (!baseURL) {
        return relativeURL || '';
    }

    if (!relativeURL) {
        return baseURL;
    }

    try {
        const absoluteURL = new URL(relativeURL, baseURL);
        return absoluteURL.href;
    } catch (error) {
        // If an error occurs (e.g., relative URL is not a valid URL), fallback to concatenating strings
        const trimmedBaseURL = baseURL.trim().replace(/\/+$/, '');
        const trimmedRelativeURL = relativeURL.trim().replace(/^\/+/, '');
        return trimmedBaseURL + '/' + trimmedRelativeURL;
    }
}
```

Reply:
> If you refer to the code below, I think it can be written more concisely.
> 
> ```ts
> function combineURLs(baseURL: string, relativeURL?: string): string {
>     if (!baseURL) {
>         return relativeURL || '';
>     }
> 
>     if (!relativeURL) {
>         return baseURL;
>     }
> 
>     try {
>         const absoluteURL = new URL(relativeURL, baseURL);
>         return absoluteURL.href;
>     } catch (error) {
>         // If an error occurs (e.g., relative URL is not a valid URL), fallback to concatenating strings
>         const trimmedBaseURL = baseURL.trim().replace(/\/+$/, '');
>         const trimmedRelativeURL = relativeURL.trim().replace(/^\/+/, '');
>         return trimmedBaseURL + '/' + trimmedRelativeURL;
>     }
> }
> ```

Reply:
it's terrible, not working function
view so simple but not readable

Reply:
` FAIL  src/common/__test__/string-builder.spec.js
  ● combineURLs should return the relative URL when the base URL have a path

    expect(received).toBe(expected) // Object.is equality

    Expected: "https://github.com/users/hg-pyun"
    Received: "https://github.com/hg-pyun"

      126 |     const result = sb.combineURLs('https://github.com/users', '/hg-pyun');
      127 |
    > 128 |     expect(result).toBe('https://github.com/users/hg-pyun');
          |                    ^
      129 | });
      130 |
      131 | test('combineURLs should return the relative URL when the base URL have a complex path', () => {

      at Object.toBe (src/common/__test__/string-builder.spec.js:128:20)

  ● combineURLs should return the relative URL when the base URL have a complex path

    expect(received).toBe(expected) // Object.is equality

    Expected: "https://github.com/foo/bar/hg-pyun"
    Received: "https://hg-pyun/"

      133 |     const result = sb.combineURLs('https://github.com/foo/bar/', '//hg-pyun');
      134 |
    > 135 |     expect(result).toBe('https://github.com/foo/bar/hg-pyun');
          |                    ^
      136 | });
      137 |

      at Object.toBe (src/common/__test__/string-builder.spec.js:135:20)

Test Suites: 1 failed, 4 passed, 5 total
Tests:       2 failed, 42 passed, 44 total
Snapshots:   0 total
Time:        1.118 s
`

Reply:
Just for reference, please use the internal logic only as a clue to simplify a bit more.